File: 01. DataStructures/03. Queue/Basic Operations/01. Queue using Array/Queue_Array.java
import java.util.*;

public class Queue_Array {

	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		System.out.println("Enter Size of Queue.");
		int size = sc.nextInt();
		Queue q = new Queue(size);

		boolean flag = true;
		int val = 0;
		while (flag) {

			System.out.println("1. Enqueue()");
			System.out.println("2. Dequeue()");
			System.out.println("3. Current Size of Queue");
			System.out.println("4. Peek()");
			System.out.println("5. View queue");
			System.out.println("6. Exit");

			System.out.println("Enter Choice");
			int choice = sc.nextInt();
			switch (choice) {

			case 1:
				System.out.println("Enter value");
				val = sc.nextInt();
				q.enqueue(val);
				break;

			case 2:
				System.out.println(q.dequeue());
				break;

			case 3:
				System.out.println(q.count());
				break;

			case 4:
				System.out.println(q.peek());
				break;

			case 5:
				q.viewQ();
				break;

			case 6:
				flag = false;
				break;

			default:
				System.out.println("Invalid Choice");
			}// switch
			System.out.println();
		} // while

	}

}

class Queue {
	private int front;
	private int rear;
	private int n;
	private int arr[];

	public Queue(int size) {
		this.n = size;
		arr = new int[n];
		front = -1;
		rear = -1;
	}

	public boolean isEmpty() {
		return (rear == -1 && front == -1);
	}

	public boolean isFull() {
		return (rear == n - 1);
	}

	public int count() {
		return rear - front + 1;
	}

	public void enqueue(int val) {
		if (isFull()) {
			System.out.println("Queue  is Full");
			return;
		}
		else if (front == -1) {
			front = rear = 0;
		}
		else {
			rear++;
		}
		arr[rear] = val;
	}

	public int dequeue() {
		int x = 0;
		if (isEmpty()) {
			System.out.println("Queue is Empty");
			return -999;
		} else if (rear == front) {
			x = arr[front];
			front = rear = -1;
			return x;
		} else {
			x = arr[front];
			front++;
			return x;

		}
	}

	public void viewQ() {
		if (isEmpty()) {
			System.out.println("Queue is empty");
			return;
		}
		for (int i = front; i <= rear; i++) {
			System.out.print(arr[i] + " ");
		}
	}

	public int peek() {
		if (isEmpty()) {
			System.out.println("Queue is empty");
			return -999;
		}
		return arr[front];
	}

}

File: 01. DataStructures/03. Queue/Basic Operations/01. Queue using Array/Queue_Array.js
// - Enqueue() – Add item to the queue from the REAR.
// - Dequeue() – Remove item from the queue from the FRONT.
// - isFull() – Check if queue is full or not.
// - isEmpty() – Check if queue empty or not.
// - count() – Get number of items in the queue.
// - peek() – return front elemenet in the queue(line).
// 1 2 3 4 5 6 7 8 9 10 
//   f     r
// [ 1,2 3,4 ]
class Queue {
    constructor(size) {
        this.queue = [];
        this.front = 0;
        this.rear = -1;
        this.size = size;
    }

    enqueue(data) {
        if(!this.isFull()) {
            this.rear++;
            this.queue.push(data);
            return this.queue;    
        }
    }

    dequeue() {
       this.queue[this.front] = null;
       this.front++;
    }

    isEmpty() {
        if(this.rear == 0) {
            return true;
        } else {
            return false;
        }
    }
    reverseQueue() {
    let left = this.front;
    let right = this.rear;

    while(left < right) {
        console.log(left, right);
        const temp = this.queue[left];
        this.queue[left] = this.queue[right];
        this.queue[right] = temp;
        left++;
        right--;
    }

    }
    count() {
        return this.queue.length - this.front;
    }

    peek()  {
        return this.queue[this.front] ? this.queue[this.front] : 'nothing is in the queue';
    }

    isFull() {
        if(this.rear < this.size - 1) {
            return false;
        }else {
            return true;
        }
    }
    print() {
        return this.queue;

    }
}


const myQueue = new Queue(5);
myQueue.enqueue(1);
myQueue.enqueue(2);
myQueue.enqueue(3);
myQueue.dequeue();
myQueue.enqueue(4);
// myQueue.enqueue(5);
// myQueue.enqueue(6);
// myQueue.enqueue(6);
myQueue.reverseQueue();
// myQueue.dequeue();
// myQueue.dequeue();

console.log(myQueue.print());
